pkg/system/system.go

```diff
@@ -0,0 +1,225 @@
+// 4. Daemons failures record as metrics
+type Controller struct {
+	manager *manager.Manager
+	// httpSever *http.Server
+	addr   *net.UnixAddr
+	router *mux.Router
+}
+
+type upgradeRequest struct {
+	NydusdPath string `json:"nydusd_path"`
+	Version    string `json:"version"`
+	Policy     string `json:"policy"`
+}
+
+type daemonInfo struct {
+	ID         string `json:"id"`
+	SnapshotID string `json:"snapshot_id"`
+	Pid        int    `json:"pid"`
+	ImageID    string `json:"image_id"`
+	APISock    string `json:"api_socket"`
+	// TODO: trim me, only for trace and analyze purpose at this stage
+	RootMountPoint   string
+	CustomMountPoint string
+	SupervisorPath   string
+	SnapshotDir      string
+}
+
+func NewSystemController(manager *manager.Manager, sock string) (*Controller, error) {
+	os.Remove(sock)
+
+	addr, err := net.ResolveUnixAddr("unix", sock)
+	if err != nil {
+		return nil, errors.Wrapf(err, "resolve address %s", sock)
+	}
+
+	sc := Controller{
+		manager: manager,
+		addr:    addr,
+		router:  mux.NewRouter(),
+	}
+
+	sc.registerRouter()
+
+	return &sc, nil
+}
+
+func (sc *Controller) registerRouter() {
+	sc.router.HandleFunc(endpointDaemons, sc.describeDaemons()).Methods(http.MethodGet)
+	sc.router.HandleFunc(endpointDaemonsUpgrade, sc.upgradeDaemons()).Methods(http.MethodPost)
+	sc.router.HandleFunc(endpointDaemonRecords, sc.getDaemonRecords()).Methods(http.MethodGet)
+
+	// Special registration for Prometheus metrics export
+	handler := promhttp.HandlerFor(exporter.Registry, promhttp.HandlerOpts{
+		ErrorHandling: promhttp.HTTPErrorOnError,
+	})
+	sc.router.Handle(endpointPromMetrics, handler)
+}
+
+func (sc *Controller) describeDaemons() func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		daemons := sc.manager.ListDaemons()
+		log.L.Infof("list daemons %v", daemons)
+
+		info := make([]daemonInfo, 0, 10)
+
+		for _, d := range daemons {
+			RootMountPointGetter := func() string {
+				if d.RootMountPoint != nil {
+					return *d.RootMountPoint
+				}
+				return ""
+			}
+
+			CustomMountPointGetter := func() string {
+				if d.CustomMountPoint != nil {
+					return *d.CustomMountPoint
+				}
+				return ""
+			}
+
+			i := daemonInfo{ID: d.ID,
+				SnapshotID:       d.SnapshotID,
+				Pid:              d.Pid,
+				ImageID:          d.ImageID,
+				RootMountPoint:   RootMountPointGetter(),
+				CustomMountPoint: CustomMountPointGetter(),
+				SnapshotDir:      d.SnapshotDir}
+
+			info = append(info, i)
+		}
+
+		w.WriteHeader(http.StatusOK)
+		w.Header().Set("Content-Type", "application/json")
+
+		respBody, err := json.Marshal(&info)
+		if err != nil {
+			panic("marshal")
+		}
+
+		w.Write(respBody)
+	}
+}
+
+// TODO: Implement me!
+func (sc *Controller) getDaemonRecords() func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		daemons := sc.manager.ListDaemons()
+		log.L.Infof("list daemons %v", daemons)
```

Comment:
Please write a `501 Not Implemented` status code at least, otherwise the API call will hang/timeout.

Reply:
fixed